pkg/ee/kubelb/controller.go

```diff
@@ -0,0 +1,458 @@
+	// Add finalizer to the cluster. We only add finalizer to the cluster if/when the kubeconfig for the
+	// kubelb management cluster is configured.
+	if !kuberneteshelper.HasFinalizer(cluster, CleanupFinalizer) {
+		if err := kuberneteshelper.TryAddFinalizer(ctx, r, cluster, CleanupFinalizer); err != nil {
+			return nil, fmt.Errorf("failed to set %q finalizer: %w", CleanupFinalizer, err)
+		}
+	}
+
+	// Create/update required resources in kubeLB management cluster.
+	if err := r.createOrUpdateKubeLBManagementClusterResources(ctx, kubeLBManagementClient, cluster, cfg); err != nil {
+		return nil, err
+	}
+
+	// Create/update required resources in user cluster.
+	if err := r.createOrUpdateKubeLBUserClusterResources(ctx, cluster); err != nil {
+		return nil, err
+	}
+
+	// Create/update required resources in user cluster namespace in seed.
+	if err := r.createOrUpdateKubeLBSeedClusterResources(ctx, cluster); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (r *reconciler) createOrUpdateKubeLBManagementClusterResources(ctx context.Context, client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster, kubeconfig []byte) error {
+	namespace := cluster.Status.NamespaceName
+
+	// Create namespace; which is equivalent to registering the tenant.
+	nsReconcilers := []reconciling.NamedNamespaceReconcilerFactory{
+		kubelbmanagementresources.NamespaceReconciler(namespace),
+	}
+
+	if err := reconciling.ReconcileNamespaces(ctx, nsReconcilers, "", client); err != nil {
+		return fmt.Errorf("failed to reconcile namespace: %w", err)
+	}
+
+	// Create RBAC for the tenant.
+	saReconcilers := []reconciling.NamedServiceAccountReconcilerFactory{
+		kubelbmanagementresources.ServiceAccountReconciler(),
+	}
+
+	if err := reconciling.ReconcileServiceAccounts(ctx, saReconcilers, namespace, client); err != nil {
+		return fmt.Errorf("failed to reconcile service account: %w", err)
+	}
+
+	roleReconcilers := []reconciling.NamedRoleReconcilerFactory{
+		kubelbmanagementresources.RoleReconciler(),
+	}
+
+	if err := reconciling.ReconcileRoles(ctx, roleReconcilers, namespace, client); err != nil {
+		return fmt.Errorf("failed to reconcile role: %w", err)
+	}
+
+	roleBindingReconcilers := []reconciling.NamedRoleBindingReconcilerFactory{
+		kubelbmanagementresources.RoleBindingReconciler(),
+	}
+
+	if err := reconciling.ReconcileRoleBindings(ctx, roleBindingReconcilers, namespace, client); err != nil {
+		return fmt.Errorf("failed to reconcile role binding: %w", err)
+	}
+
+	// Create service account token secret.
+	secretReconcilers := []reconciling.NamedSecretReconcilerFactory{
+		kubelbmanagementresources.SecretReconciler(),
+	}
+
+	if err := reconciling.ReconcileSecrets(ctx, secretReconcilers, namespace, client); err != nil {
+		return fmt.Errorf("failed to reconcile secret: %w", err)
+	}
+
+	// Generate kubeconfig secret.
+	tenantKubeconfig, err := r.generateKubeconfig(ctx, client, namespace, string(kubeconfig))
```

Comment:
shouldn't this be in `createOrUpdateKubeLBSeedClusterResources`? I think the tenant kubeconfig is used only by the `kubelb-ccm` which should run in the seed, no?